File: iosApp/iosApp/ComponentViews/StopCard/StopCardList.swift
//
//  StopCardList.swift
//  iosApp
//
//  Created by Melody Horn on 6/25/26.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct StopCardList<EmptyView: View>: View {
    @ObserveInjection var inject

    let stopCardData: [StopCardData]?
    @ViewBuilder let emptyView: () -> EmptyView
    let global: GlobalResponse?
    let now: Date
    let isFavorite: (RouteStopDirection) -> Bool
    let pushNavEntry: (SheetNavigationStackEntry) -> Void

    var body: some View {
        if let stopCardData, !stopCardData.isEmpty {
            HaloScrollView {
                LazyVStack(alignment: .center, spacing: 18) {
                    ForEach(stopCardData, id: \.stop.id) { stopCardData in
                        StopCard(
                            cardData: stopCardData,
                            global: global,
                            now: now.toEasternInstant(),
                            isFavorite: isFavorite,
                            pushNavEntry: pushNavEntry
                        )
                    }
                }
                .padding(.vertical, 4)
                .padding(.horizontal, 16)
            }
            .enableInjection()
        } else if let stopCardData, stopCardData.isEmpty {
            HaloScrollView {
                VStack {
                    emptyView()
                    Spacer()
                }
                .padding(.horizontal, 16)
            }
            .enableInjection()
        } else {
            ScrollView([]) {
                LazyVStack(alignment: .center, spacing: 14) {
                    ForEach(0 ..< 5) { _ in
                        LoadingStopCard()
                    }
                }
                .padding(.vertical, 4)
                .padding(.horizontal, 16)
                .loadingPlaceholder(withShimmer: false)
            }
            .enableInjection()
        }
    }
}

#Preview {
    let now = EasternTimeInstant.now()
    let objects = TestData.clone(namespace: "StopCardListPreview")

    let ruggles = objects.getStop(id: "place-rugg")
    let tremontAtMelneaCass = objects.stop { stop in
        stop.id = "1227"
        stop.locationType = .stop
        stop.name = "Tremont St @ Melnea Cass Blvd"
        stop.wheelchairBoarding = .accessible
    }
    let boylston = objects.getStop(id: "place-boyls")
    let ol = LineOrRoute.route(objects.getRoute(id: "Orange"))
    let olSouthboundPattern = objects.getRoutePattern(id: "Orange-3-0")
    let olNorthboundPattern = objects.getRoutePattern(id: "Orange-3-1")
    let gl = LineOrRoute.line(
        objects.getLine(id: "line-Green"),
        [
            objects.getRoute(id: "Green-B"),
            objects.getRoute(id: "Green-C"),
            objects.getRoute(id: "Green-D"),
            objects.getRoute(id: "Green-E")
        ]
    )
    let bus43 = LineOrRoute.route(objects.route { route in
        route.id = "43"
        route.color = "FFC72C"
        route.directionNames = ["Outbound", "Inbound"]
        route.directionDestinations = ["Ruggles Station", "Park Street Station"]
        route.shortName = "43"
        route.textColor = "000000"
        route.type = .bus
    })
    let bus43Inbound = objects.routePattern(route: bus43.route) { routePattern in
        routePattern.directionId = 1
        routePattern.sortOrder = 504_301_000
        routePattern.typicality = .typical
        routePattern.representativeTrip { trip in
            trip.headsign = "Park St & Tremont St"
            trip.stopIds = ["17869", "1227", "10000"]
        }
    }

    let olSouthboundLeaf = RouteCardData.Leaf(
        lineOrRoute: ol,
        stop: ruggles,
        direction: .init(directionId: 0, route: ol.route),
        routePatterns: [olSouthboundPattern],
        stopIds: Set(ruggles.childStopIds),
        upcomingTrips: [
            objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: olSouthboundPattern)
                prediction.departureTime = now.plus(minutes: 3)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: olSouthboundPattern)
                prediction.departureTime = now.plus(minutes: 11)
            })
        ],
        alertsHere: [],
        allDataLoaded: true,
        hasSchedulesToday: true,
        subwayServiceStartTime: nil,
        alertsDownstream: [objects.alert { $0.effect = .suspension }],
        context: .favorites
    )

    let olNorthboundLeaf = RouteCardData.Leaf(
        lineOrRoute: ol,
        stop: ruggles,
        direction: .init(directionId: 1, route: ol.route),
        routePatterns: [olNorthboundPattern],
        stopIds: Set(ruggles.childStopIds),
        upcomingTrips: [
            objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: olNorthboundPattern)
                prediction.departureTime = now.plus(minutes: 6)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: olNorthboundPattern)
                prediction.departureTime = now.plus(minutes: 14)
            })
        ],
        alertsHere: [],
        allDataLoaded: true,
        hasSchedulesToday: true,
        subwayServiceStartTime: nil,
        alertsDownstream: [],
        context: .favorites
    )

    let bus43RugglesLeaf = RouteCardData.Leaf(
        lineOrRoute: bus43,
        stop: ruggles,
        direction: .init(directionId: 1, route: bus43.route),
        routePatterns: [bus43Inbound],
        stopIds: Set(ruggles.childStopIds),
        upcomingTrips: [
            objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: bus43Inbound)
                prediction.departureTime = now.plus(minutes: 22)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: bus43Inbound)
                prediction.departureTime = now.plus(minutes: 56)
            }), objects.upcomingTrip(schedule: objects.schedule { schedule in
                schedule.trip = objects.trip(routePattern: bus43Inbound)
                schedule.departureTime = now.plus(hours: 12)
            })
        ],
        alertsHere: [],
        allDataLoaded: true,
        hasSchedulesToday: true,
        subwayServiceStartTime: nil,
        alertsDownstream: [],
        context: .favorites
    )

    let bus43TremontAtMelneaCassLeaf = RouteCardData.Leaf(
        lineOrRoute: bus43,
        stop: tremontAtMelneaCass,
        direction: .init(directionId: 1, route: bus43.route),
        routePatterns: [bus43Inbound],
        stopIds: [tremontAtMelneaCass.id],
        upcomingTrips: [
            objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: bus43Inbound)
                prediction.departureTime = now.plus(minutes: 25)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: bus43Inbound)
                prediction.departureTime = now.plus(minutes: 59)
            }), objects.upcomingTrip(schedule: objects.schedule { schedule in
                schedule.trip = objects.trip(routePattern: bus43Inbound)
                schedule.departureTime = now.plus(hours: 12)
            })
        ],
        alertsHere: [],
        allDataLoaded: true,
        hasSchedulesToday: true,
        subwayServiceStartTime: nil,
        alertsDownstream: [],
        context: .favorites
    )

    let glWestboundLeaf = RouteCardData.Leaf(
        lineOrRoute: gl,
        stop: boylston,
        direction: .init(name: "West", destination: nil, id: 0),
        routePatterns: [
            objects.getRoutePattern(id: "Green-B-812-0"),
            objects.getRoutePattern(id: "Green-C-832-0"),
            objects.getRoutePattern(id: "Green-D-855-0"),
            objects.getRoutePattern(id: "Green-E-886-0")
        ],
        stopIds: Set(boylston.childStopIds),
        upcomingTrips: [
            objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: objects.getRoutePattern(id: "Green-C-832-0"))
                prediction.departureTime = now.plus(minutes: 3)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: objects.getRoutePattern(id: "Green-B-812-0"))
                prediction.departureTime = now.plus(minutes: 5)
            }), objects.upcomingTrip(prediction: objects.prediction { prediction in
                prediction.trip = objects.trip(routePattern: objects.getRoutePattern(id: "Green-D-855-0"))
                prediction.departureTime = now.plus(minutes: 10)
            })
        ],
        alertsHere: [],
        allDataLoaded: true,
        hasSchedulesToday: true,
        subwayServiceStartTime: nil,
        alertsDownstream: [],
        context: .favorites
    )

    VStack {
        StopCardList(
            stopCardData: [
                .init(stop: ruggles, data: [olSouthboundLeaf, olNorthboundLeaf, bus43RugglesLeaf]),
                .init(stop: tremontAtMelneaCass, data: [bus43TremontAtMelneaCassLeaf]),
                .init(stop: boylston, data: [glWestboundLeaf])
            ],
            emptyView: {},
            global: .init(objects: objects),
            now: now.toNSDateLosingTimeZone(),
            isFavorite: { _ in true },
            pushNavEntry: { _ in }
        )
    }
    .withFixedSettings([.favoritesByStop: true, .stationAccessibility: true])
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/stopCard/StopCardList.kt
package com.mbta.tid.mbta_app.android.component.stopCard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.analytics.Analytics
import com.mbta.tid.mbta_app.analytics.MockAnalytics
import com.mbta.tid.mbta_app.android.component.ScrollSeparatorLazyColumn
import com.mbta.tid.mbta_app.android.util.IsLoadingSheetContents
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.Direction
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.model.LocationType
import com.mbta.tid.mbta_app.model.RouteCardData
import com.mbta.tid.mbta_app.model.RoutePattern
import com.mbta.tid.mbta_app.model.RouteStopDirection
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.StopCardData
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.WheelchairBoardingStatus
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.repositories.MockSettingsRepository
import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import com.mbta.tid.mbta_app.utils.TestData
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes
import org.koin.core.context.startKoin
import org.koin.core.context.stopKoin
import org.koin.dsl.module

@Composable
fun ColumnScope.StopCardList(
    stopCardData: List<StopCardData>?,
    emptyView: @Composable () -> Unit,
    global: GlobalResponse?,
    now: EasternTimeInstant,
    isFavorite: (RouteStopDirection) -> Boolean,
    onOpenStopDetails: (String, StopDetailsFilter?) -> Unit,
) {
    val contentPadding = PaddingValues(start = 15.dp, top = 7.dp, end = 15.dp, bottom = 16.dp)
    val verticalArrangement = Arrangement.spacedBy(14.dp)
    val horizontalAlignment = Alignment.CenterHorizontally
    if (stopCardData == null) {
        CompositionLocalProvider(IsLoadingSheetContents provides true) {
            ScrollSeparatorLazyColumn(
                contentPadding = contentPadding,
                verticalArrangement = verticalArrangement,
                horizontalAlignment = horizontalAlignment,
                userScrollEnabled = false,
            ) {
                items(5) { LoadingStopCard() }
            }
        }
    } else if (stopCardData.isEmpty()) {
        Column(
            Modifier.verticalScroll(rememberScrollState()).padding(8.dp).weight(1f),
            verticalArrangement = Arrangement.Center,
        ) {
            emptyView()
        }
    } else {
        ScrollSeparatorLazyColumn(
            contentPadding = contentPadding,
            verticalArrangement = verticalArrangement,
            horizontalAlignment = horizontalAlignment,
        ) {
            items(stopCardData) { StopCard(it, global, now, isFavorite, onOpenStopDetails) }
        }
    }
}

@Preview
@Composable
private fun StopCardListPreview() {
    val now = EasternTimeInstant.now()
    val objects = TestData.clone("StopCardListPreview")

    val ruggles = objects.getStop("place-rugg")
    val tremontAtMelneaCass = objects.stop {
        id = "1227"
        locationType = LocationType.STOP
        name = "Tremont St @ Melnea Cass Blvd"
        wheelchairBoarding = WheelchairBoardingStatus.ACCESSIBLE
    }
    val boylston = objects.getStop("place-boyls")
    val ol = LineOrRoute.Route(objects.getRoute("Orange"))
    val olSouthboundPattern = objects.getRoutePattern("Orange-3-0")
    val olNorthboundPattern = objects.getRoutePattern("Orange-3-1")
    val gl =
        LineOrRoute.Line(
            objects.getLine("line-Green"),
            setOf(
                objects.getRoute("Green-B"),
                objects.getRoute("Green-C"),
                objects.getRoute("Green-D"),
                objects.getRoute("Green-E"),
            ),
        )
    val bus43 =
        LineOrRoute.Route(
            objects.route {
                id = "43"
                color = "FFC72C"
                directionNames = listOf("Outbound", "Inbound")
                directionDestinations = listOf("Ruggles Station", "Park Street Station")
                shortName = "43"
                textColor = "000000"
                type = RouteType.BUS
            }
        )
    val bus43Inbound =
        objects.routePattern(bus43.route) {
            directionId = 1
            sortOrder = 504301000
            typicality = RoutePattern.Typicality.Typical
            representativeTrip {
                headsign = "Park St & Tremont St"
                stopIds = listOf("17869", "1227", "10000")
            }
        }

    val global = GlobalResponse(objects)

    val olSouthboundLeaf =
        RouteCardData.Leaf(
            ol,
            ruggles,
            Direction(0, ol.route),
            listOf(olSouthboundPattern),
            ruggles.childStopIds.toSet(),
            upcomingTrips =
                listOf(
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(olSouthboundPattern)
                            departureTime = now + 3.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(olSouthboundPattern)
                            departureTime = now + 11.minutes
                        }
                    ),
                ),
            alertsHere = emptyList(),
            allDataLoaded = true,
            hasSchedulesToday = true,
            subwayServiceStartTime = null,
            alertsDownstream = listOf(objects.alert { effect = Alert.Effect.Suspension }),
            RouteCardData.Context.Favorites,
        )

    val olNorthboundLeaf =
        RouteCardData.Leaf(
            ol,
            ruggles,
            Direction(1, ol.route),
            listOf(olNorthboundPattern),
            ruggles.childStopIds.toSet(),
            upcomingTrips =
                listOf(
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(olNorthboundPattern)
                            departureTime = now + 6.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(olNorthboundPattern)
                            departureTime = now + 14.minutes
                        }
                    ),
                ),
            alertsHere = emptyList(),
            allDataLoaded = true,
            hasSchedulesToday = true,
            subwayServiceStartTime = null,
            alertsDownstream = emptyList(),
            RouteCardData.Context.Favorites,
        )

    val bus43RugglesLeaf =
        RouteCardData.Leaf(
            bus43,
            ruggles,
            Direction(1, bus43.route),
            listOf(bus43Inbound),
            ruggles.childStopIds.toSet(),
            upcomingTrips =
                listOf(
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 22.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 56.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.schedule {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 12.hours
                        }
                    ),
                ),
            alertsHere = emptyList(),
            allDataLoaded = true,
            hasSchedulesToday = true,
            subwayServiceStartTime = null,
            alertsDownstream = emptyList(),
            RouteCardData.Context.Favorites,
        )

    val bus43TremontAtMelneaCassLeaf =
        RouteCardData.Leaf(
            bus43,
            tremontAtMelneaCass,
            Direction(1, bus43.route),
            listOf(bus43Inbound),
            setOf(tremontAtMelneaCass.id),
            upcomingTrips =
                listOf(
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 25.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 59.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.schedule {
                            trip = objects.trip(bus43Inbound)
                            departureTime = now + 12.hours
                        }
                    ),
                ),
            alertsHere = emptyList(),
            allDataLoaded = true,
            hasSchedulesToday = true,
            subwayServiceStartTime = null,
            alertsDownstream = emptyList(),
            RouteCardData.Context.Favorites,
        )

    val glWestboundLeaf =
        RouteCardData.Leaf(
            gl,
            boylston,
            Direction("West", null, 0),
            listOf(
                objects.getRoutePattern("Green-B-812-0"),
                objects.getRoutePattern("Green-C-832-0"),
                objects.getRoutePattern("Green-D-855-0"),
                objects.getRoutePattern("Green-E-886-0"),
            ),
            boylston.childStopIds.toSet(),
            upcomingTrips =
                listOf(
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(objects.getRoutePattern("Green-C-832-0"))
                            departureTime = now + 3.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(objects.getRoutePattern("Green-B-812-0"))
                            departureTime = now + 5.minutes
                        }
                    ),
                    objects.upcomingTrip(
                        objects.prediction {
                            trip = objects.trip(objects.getRoutePattern("Green-D-855-0"))
                            departureTime = now + 10.minutes
                        }
                    ),
                ),
            alertsHere = emptyList(),
            allDataLoaded = true,
            hasSchedulesToday = true,
            subwayServiceStartTime = null,
            alertsDownstream = emptyList(),
            RouteCardData.Context.Favorites,
        )

    try {
        stopKoin()
    } catch (_: Exception) {}
    startKoin {
        modules(
            module {
                single<Analytics> { MockAnalytics() }
                single {
                    SettingsCache(
                        MockSettingsRepository(
                            mapOf(
                                Settings.FavoritesByStop to true,
                                Settings.StationAccessibility to true,
                            )
                        )
                    )
                }
            }
        )
    }
    Column {
        StopCardList(
            listOf(
                StopCardData(ruggles, listOf(olSouthboundLeaf, olNorthboundLeaf, bus43RugglesLeaf)),
                StopCardData(tremontAtMelneaCass, listOf(bus43TremontAtMelneaCassLeaf)),
                StopCardData(boylston, listOf(glWestboundLeaf)),
            ),
            emptyView = {},
            global,
            now,
            isFavorite = { true },
            onOpenStopDetails = { _, _ -> },
        )
    }
}
